Add LSM6DSO support

alex-vg opened the pull request:
Hey @ginkage,

Thank you for this awesome little project!
Unfortunately, the BMI160 is currently hard to get, so I added support for the LSM6DSOTR (LSM6DSO lib).
I also got rid of the checked-in external dependencies and moved them to submodules inside the `/lib/` folder.
Additionally, I added some logic to select the correct library by scanning the I2C addresses.

I would be happy if you could merge the changes back into your repo and update the app or let me know if you have any suggestions.

The changes are only tested with LSM6DSOTR.

Cheers!

alex-vg committed:
added lsm6dso version

alex-vg committed:
moved checked in drivers to submodules

alex-vg committed:
refactored imu headers

alex-vg committed:
added dynamic lib usage depending on i2c address

ginkage commented:
Hi!
Thanks, this looks nice. I'll review and add comments as I go (sorry in advance if it will be a tad slow), but at a first glance it does look good. We'll also need to bump the app version in the Flipper store, plus I want to add a warning if the app was launched without a connected module, but all in all I'm happy to adopt this change.

ginkage commented on tracking/imu/imu_lsm6dso.h:
```diff
@@ -0,0 +1,23 @@
+
+#define LSM6DSO_TAG "LSM6DO-C"
+
+#define LSM6DSO_DEV_ADDRESS UINT8_C(0xD6)
```
Nit: for uniformity, it's probably better to rename it to `LSM6DSO_ADDRESS` and specify address as `(0x6B << 1)`.

ginkage replied:
(Actually, renaming `LSM6DS3_ADDRESS` to `LSM6DS3_DEV_ADDRESS` might be more uniform.)

ginkage commented on tracking/imu/imu_lsm6ds3trc.c:
```diff
@@ -1,14 +1,6 @@
-#define LSM6DS3_ADDRESS (0x6A << 1)
-
-static const double DEG_TO_RAD = 0.017453292519943295769236907684886;
+static const double LSM6DS3_DEG_TO_RAD = 0.017453292519943295769236907684886;
```
Nit: since the constant value and usage is the same for all three, it's probably worth moving to `imu.h` with a generic name of `DEG_TO_RAD`.

ginkage commented on tracking/imu/imu.c:
```diff
@@ -1,29 +1,85 @@
+
+imu_t* get_imu(uint8_t address) {
+    unsigned int i;
+    for(i = 0; i < 3; i++) {
```
Nit: I suggest using `sizeof()` instead of a hard-coded number.

ginkage commented on tracking/imu/imu.c:
```diff
@@ -1,29 +1,85 @@
 
-bool bmi160_begin();
-int bmi160_read(double* vec);
+imu_t imu_types[] = {
```
Listing all functions here is a bit excessive. I would suggest every header to expose just an `external struct imu_t` and define them in the corresponding .c files, then you wouldn't need to list them all twice (or define in the .h file in the first place), would make slightly better encapsulation IMHO.

ginkage replied:
(TAG names and addresses are probably better to be kept in .c files as well. Note that you can reuse TAG as a device name, too.)

ginkage commented:
Looks like all the driver files now have the `imu_` prefix, so probably worth just dropping it (that could make reviewing changes a tad trickier, but I can always just look at the specific commit with renaming).

Sorry, there are quite a few comments now, but I believe the whole thing will look significantly cleaner, and that'll make it a bit easier to add other IMUs like MPU6050 in the future.

alex-vg commented:
Thank you for your review, I will try to change it tomorrow.

ginkage commented:
No rush, and thank you once again for the PR!

alex-vg committed:
bumped version to 1.2
renamed LSM6DS3_ADDRESS to LSM6DS3_DEV_ADDRESS
moved LSM6DS3_DEG_TO_RAD into imu.h
refactored using sizeof
moved imu_t definition

alex-vg commented:
I hope I got everything.
I also added a shortcut in the IMU detection / I2C scanning. We just need to test the known addresses.

ginkage commented on tracking/imu/imu.h:
```diff
@@ -1,6 +1,13 @@
 #include <stdbool.h>
+#include <furi_hal.h>
+
+#include "imu_t.h"
+#include "bmi160.h"
+#include "lsm6ds3trc.h"
+#include "lsm6dso.h"
```
Let's delete all four of these files. :)
Because `imu_t.h` is not used anywhere else, and the only thing that's used from the other three is the `extern struct imu_t` part, so we might as well just list them here (that would still take one line per driver) and delete the headers.

ginkage commented on tracking/imu/imu.h:
```diff
@@ -9,10 +16,19 @@ extern "C" {
 #define ACC_DATA_READY (1 << 0)
 #define GYR_DATA_READY (1 << 1)
 
+#define IMU_TAG "IMU_H"
```
Nit: please keep it in the '.c' file, as it's not used anywhere else.

ginkage commented on tracking/imu/lsm6ds3trc.h:
```diff
@@ -0,0 +1,22 @@
+
+#include "imu.h"
+
+// External variable declaration
+extern stmdev_ctx_t lsm6ds3trc_ctx;
```
No need to define it here (doesn't matter though, as it's better to delete the file anyway).

ginkage commented:
Thanks, looking cleaner now!
Just a couple more comments (basically, the idea is that `.h` files define an API used by other modules, so they shouldn't contain things that aren't used in other `.c` files).

ginkage commented:
Oh well, I guess I'll clean up the rest myself. Thanks again.

alex-vg commented:
Sorry, I didn't have time to refactor it. Thank you for taking over.

ginkage commented:
No worries, and thanks once again. I've verified that my test module with LSM6DS3 works, but it actually needed some remapping (warning: the chip on that prototype is oriented in a really quirky way!). Please check that your module has axes properly oriented and do file a PR if some remapping is needed.
